### PR TITLE
Prevent invoking property 'init' method

### LINKED ISSUE #40118: Support for `init` properties in dynamic binder
See: https://github.com/dotnet/csharplang/issues/39#issuecomment-666012974

Seems that the C# runtime binder ignores `modreq` metadata on members.  This enables `init` properties to be trivially mutated by casting to `dynamic`:

```cs
public class Foo {
    public string Value { get; init; }
}

static class Program {
    static void Main() {
        var foo = new Foo { Value = "Hello" };
        ((dynamic)foo).Value = "World";
        Console.WriteLine(foo.Value); // prints World
    }
}
```

### PR DESCRIPTION
**Pull request by cston**
Prevent invoking a property setter in the dynamic binder if the setter has a `System.Runtime.CompilerServices.IsExternalInit` modifier.

Fixes #40118.

### CODE REVIEW

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Tree/Property.cs [hunk 1]
```diff
@@ -39,5 +45,11 @@ public ExprProperty(CType type, Expr pOptionalObjectThrough, Expr pOptionalArgum
         public MethWithType MethWithTypeSet { get; }
 
         public override SymWithType GetSymWithType() => PropWithTypeSlot;
+
+        internal static bool HasIsExternalInitModifier(MethWithType mwtSet)
+        {
+            var types = (mwtSet.Meth()?.AssociatedMemberInfo as MethodInfo)?.ReturnParameter.GetRequiredCustomModifiers();
+            return (types != null) && types.Any(type => type.FullName == "System.Runtime.CompilerServices.IsExternalInit");
```
Comment (jaredpar) [inline src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Tree/Property.cs:52]: Know the compiler matches by name here but do we check to make sure it is a `class` in the compiler? Or do we allow the modreq to be any kind of type?

**Review by jaredpar — APPROVED** [review]

**Review by 333fred — APPROVED** [review]
:shipit:

**Comment by RikkiGibson** [comment]
I see this is in the 5.0.0 milestone, does this need to be ported to the release/5.0 branch?

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Tree/Property.cs (hunk 1 above)
Comment (cston) [reply src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Tree/Property.cs:52]: The compiler doesn't appear to check for `class`.

The compiler does have some additional checks though. Updated here.

**Pull request merged**